File: examples/php/ds/dataset.php
<?php
namespace books;
require_once __DIR__ . '/common/base.php';
require_once __DIR__ . '/data/author.php';
require_once __DIR__ . '/data/book.php';
require_once __DIR__ . '/data/genre.php';
require_once __DIR__ . '/data/publisher.php';


class Dataset {
    public $db = null;
    public $authors = null;
    public $books = null;
    public $genres = null;
    public $publishers = null;

    function __construct(string $scheme, ?string $queries=null) {
        $this->db = new quartz\Database($scheme, $queries);
        $this->authors = new AuthorData($this->db);
        $this->books = new BookData($this->db);
        $this->genres = new GenreData($this->db);
        $this->publishers = new PublisherData($this->db);
    }

    function open(bool $reset=false) {
        $this->db->open(getenv('DB_NAME'), getenv('DB_USER'), getenv('DB_PASS'), reset:$reset);
    }

    function isOpen(): bool {
        return $this->db.isOpen();
    }
}


?>

File: examples/python/ds/dataset.py
import ds.model._dataset as _base
from ds.data.author import *
from ds.data.book import *
from ds.data.book_author import *
from ds.data.genre import *
from ds.data.publisher import *
import quartz.data.sqlite as _sqlite
import quartz.data.mysql as _mysql

#
# Dataset
#

class Dataset(_base.Dataset):

    def __init__(self, schema_path, file_path:str):
        super().__init__()
        # Database
        self._db = _sqlite.Database(schema_path, file_path)
        # self._db = _mysql.Database(os.getenv("DB_SCHEMA"), os.getenv("DB_NAME"), os.getenv("DB_USER"), os.getenv("DB_PASS"))
        # Entities
        self._authors = AuthorData(self._db)
        self._books = BookData(self._db)
        self._book_authors = BookAuthorData(self._db)
        self._genres = GenreData(self._db)
        self._publishers = PublisherData(self._db)

    @property
    def path(self):
        return self._db.path

    def open(self, reset: bool = False):
        self._db.open(reset)

    def isOpen(self):
        return self._db.isOpen()

    def close(self):
        self._db.close()

    @staticmethod
    def valid(x, nullable=False):
        if nullable and (x is None):
            return None
        if isinstance(x, Dataset):
            return x
        Error.invalid("dataset", x)
